### PR TITLE
DMPlex vertex location test

### LINKED ISSUE #9: Tests of the DMPlex creation
Make tests of the DMPlex. Determine the necessary factors to include or tolerances to respect when DMPlex cell volumes and cell centres do not match the cell volumes and cell centres of the BOUT++ cells (inferred from the BOUT++ metric).

### PR DESCRIPTION
**Pull request by mikekryjak**
A new test in `/tests/integrated/dmplex` checks that the vertices are in the same location between a Hypnotoad and a DMPlex grid for 6 Hypnotoad test grids.

Here are the steps it takes:
1. Tidy the directory (remove files from previous runs of the test)
2. Download and unzip the test grids from Zenodo
3. Figure out the number of cores to run a grid on
4. Change the input file to select a particular grid
5. Run `bout-particle-push` for a very short time to get a DMPlex created
6. Read DMPlex and Hypnotoad grids to extract vertices
7. Check that all DMPlex vertices have an equivalent in Hypnotoad to 1mm tolerance
8. Check that all Hypnotoad vertices have an equivalent in DMPlex to 1mm tolerance

### CODE REVIEW

File: tests/integrated/dmplex/runtest [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+from boutdata.data import BoutOptionsFile
+from petsc4py import PETSc
+import zipfile
+import xhermes
+
+"""
+This test checks that the DMPLex and Hypnotoad grids have matching vertices.
+It does this by:
+- Downloading a set of Hypnotoad example grids
+- Running a simple NESO-Particles example to generate the DMPlex
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: This comment seems misleading. NESO-Particles does not generate the DMPlex. The DMPlex is created by code in this repo passed to NESO-Particles.

File: tests/integrated/dmplex/runtest [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    viewer = PETSc.Viewer().createHDF5(str(dmplex_path), "r")
+
+    dm.load(viewer)
+    viewer.destroy()
+    dm.setFromOptions()
+    dm.setUp()
+
+    coords = dm.getCoordinatesLocal().array  # Get all RZ coordinates
+    section = dm.getCoordinateSection()
+    dim = dm.getCoordinateDim()  # Number of dimensions in the grid
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: Is this variable `dim` unused?

File: tests/integrated/dmplex/runtest [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    ### Check vertices
+
+    # Extract
+    dmplex_vertices = get_DMPlex_vertices(dmplex_path)
+    hypno_vertices = get_hypnotoad_vertices(gridfile)
+
+    # Ensure all DMPLex vertices exist in Hypnotoad
+    for i in range(dmplex_vertices.shape[0]):
+        R_diff = abs(hypno_vertices[:, 0] - dmplex_vertices[i, 0])/hypno_vertices[:, 0]
+        Z_diff = abs(hypno_vertices[:, 1] - dmplex_vertices[i, 1])/hypno_vertices[:, 1]
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: If a vertex has Z = 0 exactly, how would this line behave?

File: tests/integrated/dmplex/runtest [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                )
+                exit(1)
+
+    ### Check vertices
+
+    # Extract
+    dmplex_vertices = get_DMPlex_vertices(dmplex_path)
+    hypno_vertices = get_hypnotoad_vertices(gridfile)
+
+    # Ensure all DMPLex vertices exist in Hypnotoad
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: Expansion of this comment to explain how the tests below work. You create a vector which will have positive entries, and if there is a zero or small enough entry, we judge that the vertex is shared between the sets. This avoids needing to make a unique set of vertices in Hypnotoad. You should consider if the `1e-6` value used for the test can be set by the difference between duplicated vertices in the Hypnotoad test, or by the tolerance value used to sort the vertices and make in unique set in the DMPlex creation.

File: tests/integrated/dmplex/runtest [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    if len(to_remove) > 0:
+        for file in to_remove:
+            shell(f"rm {file}")
+
+    ## Download grids
+    urlretrieve(gridurl, gridzipfile)
+    with zipfile.ZipFile("Hypnotoad_examples.zip", "r") as z:
+        try:
+            z.extractall(".")
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: Is it possible to build in a check here to make sure we only extract files that we expect in advance? See the warning comment here https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.extractall.

File: tests/integrated/dmplex/runtest [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    if logfile_path.exists():
+        to_remove.append(logfile_path)
+
+    if len(to_remove) > 0:
+        for file in to_remove:
+            shell(f"rm {file}")
+
+    ## Download grids
+    urlretrieve(gridurl, gridzipfile)
```
Comment (mrhardman) [inline tests/integrated/dmplex/runtest]: This function seems to be part of the legacy interface from urllib. https://docs.python.org/3/library/urllib.request.html#legacy-interface

File: tests/integrated/dmplex-vertex-coordinates/data/BOUT.inp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+phi = 0
+ex = 0
+ey = 0
+
+[neso_particles]
+
+npart_per_cell = 1
+dt = 0.005
+nsteps = 1
```
Comment (mrhardman) [inline tests/integrated/dmplex-vertex-coordinates/data/BOUT.inp]: Does this test require any particle pushing? If not, we can set `nsteps=0`.

**Comment by mikekryjak** [comment]
I fixed the `boutdata` issue causing warnings (https://github.com/boutproject/boutdata/pull/139). I think we should probably link to `boutdata` master just like we did for xHermes and xBOUT in https://github.com/UKAEA-Edge-Code/hermes-3/pull/4, so we don't wait for people to make a release.

**Comment by mrhardman** [comment]
Can we use a specific commit or release tag? If we simply use the latest master then we are liable to see unexpected behaviour when `boutdata` is updated.

File: tests/integrated/dmplex/runtest (hunk 5 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: Done:
https://github.com/UKAEA-Edge-Code/hermes-3/blob/a60438ff665c4379a4c3b462349b6452accbf919/tests/integrated/dmplex-vertex-coordinates/runtest#L155-L157

Also added hash check:
https://github.com/UKAEA-Edge-Code/hermes-3/blob/a60438ff665c4379a4c3b462349b6452accbf919/tests/integrated/dmplex-vertex-coordinates/runtest#L162-L167

File: tests/integrated/dmplex/runtest (hunk 1 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: Fixed:
https://github.com/UKAEA-Edge-Code/hermes-3/blob/a60438ff665c4379a4c3b462349b6452accbf919/tests/integrated/dmplex-vertex-coordinates/runtest#L16-L21

File: tests/integrated/dmplex/runtest (hunk 2 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: Yes. Removed

File: tests/integrated/dmplex/runtest (hunk 3 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: Badly. Replaced the algorithm with one that calculates the error by the ratio of the distance between the points and the grid length scale. This way we avoid divide by zero issues and simplify the code.

https://github.com/UKAEA-Edge-Code/hermes-3/blob/a60438ff665c4379a4c3b462349b6452accbf919/tests/integrated/dmplex-vertex-coordinates/runtest#L214-L223

File: tests/integrated/dmplex/runtest (hunk 4 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: I expanded the comment. In terms of the Hypnotoad vertex duplicates, I think it would only be a problem in an edge case where all but one vertex are beyond the distance tolerance. Even then, having DMPlex vertices not exactly in the same place as the Hypnotoad vertices may not be a problem. All we need these cells for is to have something to get sources from and push sources to.  

New comment: 

https://github.com/UKAEA-Edge-Code/hermes-3/blob/e2186aa545067914c621d8c528c917a9809aad16/tests/integrated/dmplex-vertex-coordinates/runtest#L218-L222

File: tests/integrated/dmplex/runtest (hunk 6 above)
Comment (mikekryjak) [reply tests/integrated/dmplex/runtest]: Replaced with best practice approach which also avoids partial temp files and catches HTTP errors:

https://github.com/UKAEA-Edge-Code/hermes-3/blob/e2186aa545067914c621d8c528c917a9809aad16/tests/integrated/dmplex-vertex-coordinates/runtest#L137-L148

File: tests/integrated/dmplex-vertex-coordinates/data/BOUT.inp (hunk 7 above)
Comment (mikekryjak) [reply tests/integrated/dmplex-vertex-coordinates/data/BOUT.inp]: Reduced to 0

**Comment by mikekryjak** [comment]
> Can we use a specific commit or release tag? If we simply use the latest master then we are liable to see unexpected behaviour when `boutdata` is updated.

Ideally, yes... in https://github.com/UKAEA-Edge-Code/hermes-3/pull/4

File: tests/integrated/dmplex-vertex-coordinates/runtest [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    hypno_vertices = get_hypnotoad_vertices(gridfile)
+    
+    R_dmplex = dmplex_vertices[:, 0]
+    Z_dmplex = dmplex_vertices[:, 1]
+    R_hypno = hypno_vertices[:, 0]
+    Z_hypno = hypno_vertices[:, 1]
+
+    # Vertex position error tolerance is factor of characteristic geometry size
+    geometry_size = np.max(np.sqrt(R_hypno**2 + Z_hypno**2))
+    tolerance = 1e-6 * geometry_size
```
Comment (mrhardman) [inline tests/integrated/dmplex-vertex-coordinates/runtest]: I still think it is important to use a value for this tolerance which is correlated with the numerical accuracy in the hypnotoad equilibrium generation and the DMPlex creation. Is it possible to determine the error between duplicated Hypnotoad vertex points and use this as the tolerance here, and also in the DMPlex creation in `bout-particle-push`? The tolerance can be read in the BOUT.inp file into a function in `bout-particle-push`.

**Review by mrhardman — CHANGES_REQUESTED** [review]
Can you please merge VANTAGE-Reactions-Coupling/dev into this branch to trigger CI tests? Is it possible to make this DMPlex creation test run as part of the CI test suite?

**Comment by mikekryjak** [comment]
> Can you please merge VANTAGE-Reactions-Coupling/dev into this branch to trigger CI tests? Is it possible to make this DMPlex creation test run as part of the CI test suite?

If I understood our last discussion correctly, we are unable to do tests in CI involving any parts of VANTAGE because we don't have a Spack install in CI yet. This will require a docker image and be quite a bit of work. I think that needs to be a separate PR later on.

File: tests/integrated/dmplex-vertex-coordinates/runtest (hunk 8 above)
Comment (mikekryjak) [reply tests/integrated/dmplex-vertex-coordinates/runtest]: This is now complete. Let me know if you're happy to merge

**Comment by mrhardman** [comment]
> > Can you please merge VANTAGE-Reactions-Coupling/dev into this branch to trigger CI tests? Is it possible to make this DMPlex creation test run as part of the CI test suite?
> 
> If I understood our last discussion correctly, we are unable to do tests in CI involving any parts of VANTAGE because we don't have a Spack install in CI yet. This will require a docker image and be quite a bit of work. I think that needs to be a separate PR later on.

I would prefer if you would add the test to this block https://github.com/UKAEA-Edge-Code/hermes-3/blob/2470142d04142dc276516d0151f7e61d6549907f/CMakeLists.txt#L417-L423, so that it is included in the integrated tests if `HERMES_USE_VANTAGE=ON`.

The docker image is almost ready.

**Comment by mikekryjak** [comment]
@mrhardman thanks for the suggestion, I hadn't seen that bit of `CMakeLists` before. I just added it. I also found that I must have made a typo on `Reactions` in .gitmodules in the `add-develop-xhermes` PR.  the indentation was wrong and it was missing `ignore = dirty`. I have merged dev into this branch and fixed that as part of the merge.

**Comment by mikekryjak** [comment]
@mrhardman I think I addressed all your comments.

File: tests/integrated/dmplex-vertex-coordinates/runtest (hunk 8 above)
Comment (mrhardman) [reply tests/integrated/dmplex-vertex-coordinates/runtest]: So if I understand correctly, you use 1.0e-8 in the BOUT.inp and 1.0e-6 = 100*1.0e-8 in the python test? Is it possible to link these two numbers explicitly? See the code here to see how I made BOUT.inp files at runtime so that the parameters therein can be controlled from the python script. https://github.com/boutproject/hermes-3/blob/920ba829cc78cdab0dbf6101c69fecc4689bd8dd/tests/mms_operator/job_functions.py#L42-L99

Comment (mikekryjak) [reply tests/integrated/dmplex-vertex-coordinates/runtest]: Yes, very easily - just done.

**Review by mrhardman — APPROVED** [review]
